driver-core/src/main/com/mongodb/client/model/expressions/Expression.java

```diff
@@ -97,4 +97,25 @@ public interface Expression {
      * @return true if less than or equal to, false otherwise
      */
     BooleanExpression lte(Expression lte);
+
+    /**
+     * also checks for nulls
+     * @param or
+     * @return
+     */
+    BooleanExpression isBooleanOr(BooleanExpression or);
+    NumberExpression isNumberOr(NumberExpression or);
+    StringExpression isStringOr(StringExpression or);
+    DateExpression isDateOr(DateExpression or);
+    ArrayExpression<Expression> isArrayOr(ArrayExpression<? extends Expression> or);
```

Comment:
Discussed, and this is correct, because all array expressions (and expressions in general) are immutable (there are no methods where the consequences of the unchecked cast would be observable.

Reply:
Will extend this to concat and union.